optimum/intel/neural_compressor/modeling_decoder.py

```diff
@@ -0,0 +1,299 @@
+logger = logging.getLogger(__name__)
+
+
+TEXT_GENERATION_EXAMPLE = r"""
+    Example of text generation:
+    ```python
+    >>> from transformers import {processor_class}
+    >>> from optimum.intel import {model_class}
+
+    >>> tokenizer = {processor_class}.from_pretrained("{checkpoint}")
+    >>> model = {model_class}.from_pretrained("{checkpoint}", export=True)
+    >>> inputs = tokenizer("I love this story because", return_tensors="pt")
+    >>> gen_tokens = model.generate(**inputs, do_sample=True, temperature=0.9, min_length=20, max_length=20)
+    >>> tokenizer.batch_decode(gen_tokens)
+    ```
+    Example using `transformers.pipelines`:
+    ```python
+    >>> from transformers import {processor_class}, pipeline
+    >>> from optimum.intel import {model_class}
+
+    >>> tokenizer = {processor_class}.from_pretrained("{checkpoint}")
+    >>> model = {model_class}.from_pretrained("{checkpoint}", export=True)
+    >>> gen_pipeline = pipeline("text-generation", model=model, tokenizer=tokenizer)
+    >>> text = "I love this story because"
+    >>> gen = gen_pipeline(text)
+    ```
+"""
+
+
+@add_start_docstrings(
+    """
+    Base INCBaseDecoderModel class.
+    """,
+)
+class INCBaseDecoderModel(INCBaseModel):
+    main_input_name = "input_ids"
+
+    def __init__(
+        self,
+        model,
+        config: PretrainedConfig = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        use_cache: bool = True,
+        **kwargs,
+    ):
+        super().__init__(
+            model,
+            config,
+            model_save_dir=model_save_dir,
+            **kwargs,
+        )
+
+        self.use_cache = use_cache
+        self.normalized_config = NormalizedConfigManager.get_normalized_config_class(config.model_type)(config)
+
+
+@add_start_docstrings(
+    """
+    Neural-compressor Model with a causal language modeling head on top (linear layer with weights tied to the input
+    embeddings).
+    """,
+    MODEL_START_DOCSTRING,
+)
+class INCModelForCausalLM(INCBaseDecoderModel, GenerationMixin):
+    export_feature = "text-generation"
+    auto_model_class = AutoModelForCausalLM
+
+    @add_start_docstrings_to_model_forward(
+        INPUTS_DOCSTRING.format("batch_size, sequence_length")
+        + TEXT_GENERATION_EXAMPLE.format(
+            processor_class=_TOKENIZER_FOR_DOC,
+            model_class="INCModelForCausalLM",
+            checkpoint="EleutherAI/gpt-j-6b",
+        )
+    )
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.FloatTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.Tensor]]] = None,
+        **kwargs,
+    ) -> CausalLMOutputWithPast:
+        inputs = {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+        }
+
+        nb_pkv = 2
+        num_layers = self.normalized_config.num_layers
+        first_token = False
+        if self.use_cache:
```

Comment:
Already integrated so should be reused : https://github.com/huggingface/optimum-intel/blob/c48f06ad3303df8f1c530db6f953c23f6e71553c/optimum/intel/generation/modeling.py#L135

Reply:
This forward function support eninge model, there is preproccess for inputs to adapt the engine model.  If we can add the preproccess in TSModelForCausalLM class, then we can remove formard function.

Reply:
it would be better than having duplicate yes

Reply:
Done